ovn-tester/cms/openstack/openstack.py

```diff
@@ -0,0 +1,724 @@
+                        due to the port naming convention and limits on
+                        interface name length, this name can't be longer than
+                        12 characters.
+        :return: LSPort representing VM's network interface
+        """
+        port_name = f"lp-{vm_name}"
+        if len(port_name) > 15:
+            raise RuntimeError(
+                f"Maximum port name length is 15. Port {port_name} is too "
+                f"long. Consider using shorter VM name."
+            )
+        port_addr = neutron_net.next_host_ip()
+        net_addr: netaddr.IPNetwork = neutron_net.network.cidr
+        port_ext_ids = {
+            "neutron:cidrs": f"{port_addr}/{net_addr.prefixlen}",
+            "neutron:device_id": str(uuid.uuid4()),
+            "neutron:device_owner": "compute:nova",
+            "neutron:network_name": neutron_net.name,
+            "neutron:port_name": "",
+            "neutron:project_id": project_id,
+            "neutron:revision_number": "1",
+            "neutron:security_group_ids": neutron_net.security_group.name,
+            "neutron:subnet_pool_addr_scope4": "",
+            "neutron:subnet_pool_addr_scope6": "",
+        }
+        port_options = (
+            f"mcast_flood_reports=true "
+            f"requested-chassis={chassis.container}"
+        )
+        ls_port = self.nbctl.ls_port_add(
+            lswitch=neutron_net.network,
+            name=port_name,
+            mac=str(RandMac()),
+            ip=DualStackIP(port_addr, net_addr.prefixlen, None, None),
+            ext_ids=port_ext_ids,
+            gw=DualStackIP(neutron_net.gateway, None, None, None),
+            metadata=chassis,
+        )
+        self.nbctl.ls_port_set_set_options(ls_port, port_options)
+        self.nbctl.ls_port_enable(ls_port)
+        self.nbctl.port_group_add_ports(
+            pg=neutron_net.security_group, lports=[ls_port]
+        )
+
+        return ls_port
+
+    def _add_metadata_port(self, network: LSwitch, project_id: str) -> LSPort:
+        """Create metadata port in LSwitch with Neutron's external IDs."""
+        port_ext_ids = {
+            "neutron:cidrs": "",
+            "neutron:device_id": f"ovnmeta-{network.uuid}",
+            "neutron:device_owner": "network:distributed",
+            "neutron:network_name": network.name,
+            "neutron:port_name": "",
+            "neutron:project_id": project_id,
+            "neutron:revision_number": "1",
+            "neutron:security_group_ids": "",
+            "neutron:subnet_pool_addr_scope4": "",
+            "neutron:subnet_pool_addr_scope6": "",
+        }
+
+        port = self.nbctl.ls_port_add(
+            lswitch=network,
+            name=str(uuid.uuid4()),
+            mac=str(RandMac()),
+            ext_ids=port_ext_ids,
+        )
+        self.nbctl.ls_port_set_set_type(port, "localport")
+        self.nbctl.ls_port_set_set_options(port, "requested-chassis=")
```

Comment:
I will remove the empty chassis key in ml2/ovn, it's not needed: https://review.opendev.org/c/openstack/neutron/+/897489